apiv3: User (Un-)Locking functionality

oliverguenther opened the pull request:
This commit implements the proposed user locking/unlocking functionality
in https://github.com/opf/openproject/pull/2401.

The APIv3 is extended with a resource `/user/:id/lock` with the following properties:
- `POST /lock`: Set the lock of user(:id), if its account status is 'active'.
- `DELETE /lock`: Remove an active lock of user(:id), if its account status is 'locked'.

This commit also implements the following changes to `UserRepresenter`:
- Replaced the integer status field with the textual status name (i.e., 'active')
- Added two entries to __links_ for admin users: `lock`, `unlock`.

The work package containing this user story: https://community.openproject.org/work_packages/18122

oliverguenther committed:
apiv3: User (Un-)Locking functionality

This commit implements the proposed user locking/unlocking functionality
in https://github.com/opf/openproject/pull/2401.

The APIv3 is extended with a resource ``/user/:id/lock`` with the following properties:

* ``POST /lock``: Set the lock of user(:id), if its account status is 'active'.

* ``DELETE /lock``: Remove an active lock of user(:id), if its account status is 'locked'.

This commit also implements the following changes to ``UserRepresenter``:

* Replaced the integer status field with the textual status name (i.e., 'active')
* Added two entries to *_links* for admin users: ``lock``, ``unlock``.

TeatroIO commented:
I've prepared a stage. <a href='http://oliverguenther-feature-apiv3-user-lockunlock.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

oliverguenther committed:
houndci linting fixes

oliverguenther commented on lib/api/v3/users/user_representer.rb:
```diff
@@ -58,6 +58,22 @@ def initialize(model, options = {}, *expand)
+            href: api_v3_paths.user_lock(represented.id),
+            title: "Remove lock on #{represented.login}",
+            method: :delete
+          } if current_user_is_admin && represented.status == User::STATUSES[:locked]
```
According to the specs, `represented`can also be a Group instance, which is why the--more readable --members `active?`, `locked?` are not available.

Also, as the `/lock` resource is accessible from admins only, their links should also be hidden from regular users.

ulferts replied:
How about adding a `lockable?(user)` method to the user/principal/group model to handle this on a more appropriate level. The group could then always respond with false, the user could respond based on the logic shown here.

Same should apply for `activatable?(user)`.

I guess not only the API would benefit from this.

ulferts commented on lib/api/v3/users/user_representer.rb:
```diff
@@ -78,12 +94,16 @@ def initialize(model, options = {}, *expand)
         property :created_at, getter: -> (*) { created_on.utc.iso8601 }, render_nil: true
         property :updated_at, getter: -> (*) { updated_on.utc.iso8601 }, render_nil: true
-        property :status, getter: -> (*) { status }, render_nil: true
+        property :status, getter: -> (*) { status_name }, render_nil: true
```
well spotted :+1:

ulferts commented on spec/lib/api/v3/users/user_representer_spec.rb:
```diff
@@ -55,10 +57,38 @@
+      end
+
+      context 'when current_user is admin' do
+        let(:current_user) { FactoryGirl.create(:admin) }
```
Whenever possible, I would like to have build_stubbed/build used instead of create. At least the tests of the higher layers (view, controller) should strive for it. Using it speeds up the tests. It might not matter for one test but doing that on a lot of them does.

ulferts commented on spec/lib/api/v3/users/user_representer_spec.rb:
```diff
@@ -55,10 +57,38 @@
+
+        context 'when account is locked' do
+          it 'should link to unlock' do
+            user.lock
```
that one could be mocked as well.

ulferts commented on spec/requests/api/v3/user/userlock_resource_spec.rb:
```diff
@@ -0,0 +1,86 @@
+
+    before do
+      allow(User).to receive(:current).and_return current_user
+      allow(User).to receive(:lock!)
```
That could be removed IMO.

ulferts commented on spec/requests/api/v3/user/userlock_resource_spec.rb:
```diff
@@ -0,0 +1,86 @@
+  let(:representer) { ::API::V3::Users::UserRepresenter.new(model) }
+  let(:lock_path) { "/api/v3/users/#{user.id}/lock" }
+
+  describe '#post' do
```
There does not seem to be a tests for DELETE.

ulferts commented on lib/api/v3/users/users_api.rb:
```diff
@@ -52,6 +52,42 @@ class UsersAPI < Grape::API
+                else
+                  fail ::API::Errors::InvalidUserStatusTransition
+                end
+              end
```
Except for `@user.active!` vs `@user.lock!` the two methods are basically the same. So this should be refactored to avoid duplication.

ulferts replied:
The `current_user` is not provided for the representer. Therefore, the links for locking/unlocking will always be missing in the response returned from those two endpoints. 

Changing this to

`UserRepresenter.new(@user, current_user: curren_user)`

Will help here. You do it correctly in the representer specs.

The GET action for user is also missing the current_user. Due to that, the link for locking/unlocking is currently displayed nowhere.

Additionally, but arguably outside of the scope of this PR, it should be added for `available_(assignees|responsibles)` as well.

ulferts commented:
@oliverguenther thank you very much for the PR. I tested it and it looks good functionality wise. Please see the comments attached to the code for what I think should be improved before merging.

oliverguenther committed:
Updated API for user (un-)locking

- Adds an lockable?, activatable? to the Principal model,
which is overridden by User as aliases to active? and locked?.
- UserRepresenter refers to the above methods
- Refactored /lock routes into a single method

Specs:
- Adds delete /users/:id/lock spec
- Refactored FactoryGirl calls to avoid create where feasible

oliverguenther commented:
Hi Jens, thanks for your comments - any remarks to reduce your work now + in further PRs is greatly appreciated.

To summarize the changes:
- Added an lockable?, activatable? to the Principal model,
  which is overridden by User as aliases to active? and locked?.
- UserRepresenter refers to the above methods
- Refactored /lock routes into a single method

Specs:
- Adds delete /users/:id/lock spec
- Refactored FactoryGirl calls to avoid create where feasible. Does it make sense to stub users for calls around the API? (c.f., the remaining `.create` calls for the users under test).
- Concerning your remark under `user.lock` (spec/lib/api/v3/users/user_representer_spec.rb:87) - Do you mean the lock statement (which is nothing more than a local status change?) or was that one related to FactoryGirl, as well?

ulferts commented:
Hi Oliver,

thanks for the update. The PR looks good to me now. @NobodysNightmare wanted to take a look at the PR as well so I refrain from merging until he had the chance.

It would be very nice of you if you where to fix hound's remarks.

Regarding the spec questions: I once read a book (You can spot great sentences by them having such a beginning) stating that unit tests should have no external dependency and that includes accessing the DB. I don't think that feasible e.g. for unit tests of the model layer but it is still a good rule of thumb.  

As representer specs are unit tests I would prefer not having DB-interactions there. So `create` statements should be avoided there. I mistook `lock` for `lock!` so please forget my comment regarding that one.

Request specs on the other hand are integrative tests and as such necessarily access the DB. Because of that however, I always strive to reduce them to the bare minimum and instead have more unit tests. But apart from the mocking we do with regards to `User.current`, I wouldn't want any stubbing in integrative tests.

So again, well done and thank you.

NobodysNightmare commented:
Except for what hound has to say,  this looks good to me, too :-)

![A picture of @houndci](http://upload.wikimedia.org/wikipedia/commons/3/30/Bloodhound_2.jpg)

oliverguenther committed:
Hound linting

oliverguenther commented:
Fixed the linting errors (didn't know that hound doesn't do much other than running rubocop.) :sweat_smile:

ulferts commented:
:+1: Thanks again @oliverguenther
